src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java

```diff
@@ -1032,6 +1032,10 @@ private boolean _checkIfCreatorPropertyBased(BeanDescription beanDesc,
                 }
             }
         }
+        // [databind#3897]: Record canonical constructor will have implicitly named propDef
+        if (propDef != null && !propDef.isExplicitlyNamed() && beanDesc.isRecordType()) {
```

Comment:
Minor thing but can remove first check and move inside previous if-block (which checks for `propDef != null`).

Reply:
Actually I agonized over these 2 options:
<table>
  <thead>
  <tr>
    <th>Option 1</th>
    <th>Option 2</th>
  </tr>
  </thead>
  <tbody>
    <tr>
        <td>
<pre>
private boolean _checkIfCreatorPropertyBased(...) {
    ...
    if (propDef != null) {
        ...
        if (implName != null && !implName.isEmpty()) {
            if (propDef.couldSerialize()) {
                return true;
            } else if (beanDesc.isRecordType()) {
                // [databind#3897]: Record canonical constructor will have implicitly named propDef
                return true;
            }
        }
    }
    // in absence of everything else, default to delegating
    return false;
}
</pre>
        </td>
        <td>
<pre>
private boolean _checkIfCreatorPropertyBased(...) {
    ...
    if (propDef != null) {
        ...
    }
    // [databind#3897]: Record canonical constructor will have implicitly named propDef
    if (propDef != null && !propDef.isExplicitlyNamed() && beanDesc.isRecordType()) {
        return true;
    }
    // in absence of everything else, default to delegating
    return false;
}
</pre>
        </td>
    </tr>
  </tbody>
</table>

I chose Option 2 because it looks easier to understand with just a quick glance.  But I'll change it if you feel strongly about Option 1.